File: 763-special-binary-string/special-binary-string.cpp
// Brute Force Code & Optimal Code
class Solution {
public:
    string makeLargestSpecial(string s) {
        // list to store all top-level special substrings
        vector<string> specials;
        // starting index of the current special substring
        int start = 0;
        // balance counter: '1' increases balance &  '0' decreases balance
        int balance = 0;
        // traverse the string to split it into top-level special substrings
        for (int i = 0; i < s.length(); i++) {
            // update the balance
            if (s[i] == '1') {
                balance++;
            } else {
                balance--;
            }
            // when balance becomes zero a complete special substring is found
            if (balance == 0) {
                // extract the inner portion by removing the outermost '1' and '0'
                string inner = s.substr(start + 1, i - start - 1);
                // recursively rearrange the inner substring to make it lexicographically largest
                string largestInner = makeLargestSpecial(inner);
                // rebuild the special substring
                string special = "1" + largestInner + "0";
                // store the special substring
                specials.push_back(special);
                // start searching for the next top-level special substring
                start = i + 1;
            }
        }
        // arrange all top-level special substrings in descending lexicographical order
        sort(specials.rbegin(), specials.rend());
        // build the final largest special string
        string result = "";
        for (string current : specials) {
            result += current;
        }
        // return the lexicographically largest special binary string
        return result;
    }
};

// Time Complexity : O(N^2logN)
// Space Complexity : O(N)

File: 763-special-binary-string/special-binary-string.py
# Brute Force Code & Optimal Code
class Solution:
    def makeLargestSpecial(self, s: str) -> str:
        # list to store all top-level special substrings
        specials = []
        # starting index of the current special substring
        start = 0
        # balance counter: +1 for '1', -1 for '0'
        balance = 0
        # traverse the string to split it into top-level special substrings
        for i in range(len(s)):
            # update the balance '1' increases it, '0' decreases it
            balance += 1 if s[i] == '1' else -1
            # when the balance becomes zero a complete special substring is found
            if balance == 0:
                # extract the inner portion by removing the outermost '1' and '0'
                inner = s[start + 1:i]
                # recursively rearrange the inner substring to make it lexicographically largest then rebuild the special substring
                specials.append("1" + self.makeLargestSpecial(inner) + "0")
                # start searching for the next top-level special substring
                start = i + 1
        # arrange all top-level special substrings in descending lexicographical order
        specials.sort(reverse=True)
        # build the final largest special string
        result = ""
        for string in specials:
            result += string
        # return the lexicographically largest special binary string
        return result

# Time Complexity : O(n^2logN)
# Space Complexity : O(N)